File: lab/1/gradient_test.py
import math
from tensor import Tensor

# y = a * x * x + b * x + c
# by hande
a = 3.1
x = 2.0
b = -1.2
c = 99.0
y = a * x * x + b * x + c
dy_da = x * x
dy_dx = 2.0 * a * x + b
dy_db = x
dy_dc = 1.0
print(f'by hand:     y={y}, dy_da={dy_da}, dy_dx={dy_dx}, dy_db={dy_db}, dy_dc={dy_dc}')

# by autograd
A = Tensor(a)
X = Tensor(x)
B = Tensor(b)
C = Tensor(c)
Y = A * X * X + B * X + C
Y.backward()
dY_dA = A.gradient
dY_dX = X.gradient
dY_dB = B.gradient
dY_dC = C.gradient
print(f'by autograd: Y={Y.value}, dY_dA={dY_dA}, dY_dX={dY_dX}, dY_dB={dY_dB}, dY_dC={dY_dC}')

assert math.isclose(y, Y.value) # Python 3.5 or newer
assert math.isclose(dy_da, dY_dA)
assert math.isclose(dy_dx, dY_dX)
assert math.isclose(dy_db, dY_dB)
assert math.isclose(dy_dc, dY_dC)

File: lab/2/gradient_test.cc
#include "lab/2/tensor.h"
#include <gtest/gtest.h>

using zeronn::lab2::Tensor;

TEST(Tensor, Gradient) {
  // y = a * x * x + b * x + c
  // by hand
  float a = 3.1;
  float x = 2.0;
  float b = -1.2;
  float c = 99.0;
  float y = a * x * x + b * x + c;
  float dy_da = x * x;
  float dy_dx = 2.0 * a * x + b;
  float dy_db = x;
  float dy_dc = 1.0;
  // by autograd
  Tensor A(a);
  Tensor X(x);
  Tensor B(b);
  Tensor C(c);
  auto Y = A * X * X + B * X + C;
  EXPECT_EQ(Y.data_, y);
  Y.Backward();
  EXPECT_EQ(A.gradient_, dy_da);
  EXPECT_EQ(X.gradient_, dy_dx);
  EXPECT_EQ(B.gradient_, dy_db);
  EXPECT_EQ(C.gradient_, dy_dc);
  // reset gradient
  Y.ResetGradient();
  Y.Backward();
  EXPECT_EQ(A.gradient_, dy_da);
  EXPECT_EQ(X.gradient_, dy_dx);
  EXPECT_EQ(B.gradient_, dy_db);
  EXPECT_EQ(C.gradient_, dy_dc);
  // release intermediate tensors
  EXPECT_EQ(Tensor::ReleaseIntermediateTensors(), 5);

}
